src/api/routes/v1/get-network.ts

```diff
@@ -0,0 +1,202 @@
+    }
+
+    /**
+     * Helper method when a promise is rejected.
+     *
+     * @param error - The error.
+     */
+    function onError(error: Error): void {
+      // skip if already resolved
+      if (resolved) {
+        return
+      }
+
+      // collect error
+      errors = errors.concat(error)
+
+      // reject promise combinator if all promises are failed/undefined
+      if (undefinedValues + errors.length === promises.length) {
+        reject(errors)
+      }
+    }
+
+    promises.forEach(async (promise) => promise.then(onFulfill, onError))
+  })
+}
+
+/**
+ * Fetches the crawl data for the node.
+ *
+ * @param host - The host URL to crawl.
+ * @returns The crawl data from the node.
+ * @throws If none of the possible crawl ports work.
+ */
+async function fetchCrawls(host: string): Promise<Crawl> {
+  const promises: Array<Promise<Crawl | undefined>> = []
+  for (const port of CRAWL_PORTS) {
+    promises.push(crawlNode(host, port))
+  }
+  return any(promises)
+}
+
+/**
+ * Checks whether the UNL of the node is recorded in the networks table.
+ *
+ * @param unl - The UNL of the node.
+ * @returns Whether the UNL of the node has been seen before.
+ */
+async function getNetworkFromUNL(unl: string): Promise<string | undefined> {
+  const result = await query('networks')
+    .select('network')
+    .where('unls', 'like', `%${unl}%`)
+  return result.length > 0 ? result[0].network : undefined
+}
+
+/**
+ * Checks whether the public key of the node is recorded in the crawls table.
+ *
+ * @param publicKey - The public key of the node.
+ * @returns Whether the public key of the node has been seen before.
+ */
+async function getNetworkFromPublicKey(
+  publicKey: string,
+): Promise<string | undefined> {
+  const result = await query('crawls')
+    .select('public_key', 'networks')
+    .where('public_key', '=', publicKey)
+  return result.length > 0 ? result[0].networks : undefined
+}
+
+/**
+ * Adds the node (and its corresponding network) to the networks table.
+ *
+ * @param url - The URL endpoint of the node.
+ * @param unl - The UNL of the node.
+ * @returns The ID of the new network.
+ */
+async function addNode(url: string, unl: string | null): Promise<string> {
+  const currentNetworks = await query('networks')
+    .select('network')
+    .orderBy('network')
+  const maxNetwork =
+    currentNetworks[currentNetworks.length - networks.length - 1]?.network ?? 0
+  const newNetwork = (Number(maxNetwork) + 1).toString()
+  const network: Network = {
+    network: newNetwork,
+    entry: url,
+    port: 51235,
+    unls: unl ? [unl] : [],
+  }
+  await query('networks').insert({
+    ...network,
+    unls: network.unls.join(','),
+  })
+
+  const crawler = new Crawler()
+  void crawler.crawl(network)
+
+  return newNetwork
+}
+
+/**
+ * Add entry to new network.
+ *
+ * @param req - Express request.
+ * @param res - Express response.
+ * @returns The Express response.
+ */
+export default async function getNetwork(
```

Comment:
I wouldn't expect `getNetwork` to modify the network - Maybe `putNetwork` would be a better name?

Reply:
(Or maybe the description for this function should make it sound more like it's caching the result instead of adding a new network)

Reply:
That's fair, only name I could think of is `fetchNetwork`, but that's not much better - just the description update is clear enough for me :)

Reply:
What about something like `getNetworkOrAdd`?

IMO the method name isn't that important, the API endpoint matters more

Reply:
I think `getNetworkOrAdd` conveys the functionality more clearly